pkg/services/ngalert/sender/dispatcher.go

```diff
@@ -0,0 +1,280 @@
+package sender
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+)
+
+// Dispatcher is a service that handles alert notifications. It accepts an alert that is represented by models.AlertRuleKey and state.State,
+// then converts it to a format that is recognizable by Alertmanager and routes the resulting alert to either an external AlertManager or notifier.MultiOrgAlertmanager or both.
+// In order to maintain the internal state, the service should be Run after creation. This method starts synchronization that is run regularly.
+type Dispatcher struct {
+	AdminConfigMtx   sync.RWMutex
+	logger           log.Logger
+	clock            clock.Clock
+	adminConfigStore store.AdminConfigurationStore
+
+	// Senders help us send alerts to external Alertmanagers.
+	Senders          map[int64]*Sender
+	SendersCfgHash   map[int64]string
+	MultiOrgNotifier *notifier.MultiOrgAlertmanager
+	SendAlertsTo     map[int64]models.AlertmanagersChoice
+
+	appURL                  *url.URL
+	disabledOrgs            map[int64]struct{}
+	adminConfigPollInterval time.Duration
+}
+
+func NewDispatcher(multiOrgNotifier *notifier.MultiOrgAlertmanager, store store.AdminConfigurationStore, clk clock.Clock, appURL *url.URL, disabledOrgs map[int64]struct{}, configPollInterval time.Duration) *Dispatcher {
+	d := &Dispatcher{
+		AdminConfigMtx:   sync.RWMutex{},
+		logger:           log.New("ngalert-notifications-dispatcher"),
+		clock:            clk,
+		adminConfigStore: store,
+
+		Senders:          map[int64]*Sender{},
+		SendersCfgHash:   map[int64]string{},
+		MultiOrgNotifier: multiOrgNotifier,
+		SendAlertsTo:     map[int64]models.AlertmanagersChoice{},
+
+		appURL:                  appURL,
+		disabledOrgs:            disabledOrgs,
+		adminConfigPollInterval: configPollInterval,
+	}
+	return d
+}
+
+func (d *Dispatcher) adminConfigSync(ctx context.Context) error {
```

Comment:
If you don't want to synchronize in the constructor, you might want to lead this function off with a call to `SyncAndApplyConfigFromDatabase`. Otherwise there will be an initial period of 60s with no configuration on the Dispatcher, right?

edit: I'm leaving this comment here because I think it illustrates my point about requiring the user to call SyncAndApply themselves being confusing.